tests/components/vizio/test_media_player.py

```diff
@@ -0,0 +1,253 @@
+            domain=DOMAIN, data=MOCK_SPEAKER_CONFIG, unique_id=UNIQUE_ID
+        )
+    else:
+        vizio_device_class = VIZIO_DEVICE_CLASS_TV
+        config_entry = MockConfigEntry(
+            domain=DOMAIN, data=MOCK_USER_VALID_TV_CONFIG, unique_id=UNIQUE_ID
+        )
+
+    with patch(
+        "homeassistant.components.vizio.media_player.VizioAsync.get_current_volume",
+        return_value=int(MAX_VOLUME[vizio_device_class] / 2),
+    ), patch(
+        "homeassistant.components.vizio.media_player.VizioAsync.get_power_state",
+        return_value=vizio_power_state,
+    ):
+        await hass.config_entries.async_add(config_entry)
+        await hass.async_block_till_done()
+
+        attr = hass.states.get(ENTITY_ID).attributes
+        assert attr["friendly_name"] == NAME
+        assert attr["device_class"] == ha_device_class
+
+        assert hass.states.get(ENTITY_ID).state == ha_power_state
+        if ha_power_state == STATE_ON:
+            assert attr["source_list"] == INPUT_LIST
+            assert attr["source"] == CURRENT_INPUT
+            assert (
+                attr["volume_level"]
+                == float(int(MAX_VOLUME[vizio_device_class] / 2))
+                / MAX_VOLUME[vizio_device_class]
+            )
+
+
+async def _test_setup_failure(hass: HomeAssistantType, config: str) -> None:
+    """Test generic Vizio entity setup failure."""
+    with patch(
+        "homeassistant.components.vizio.media_player.VizioAsync.can_connect",
+        return_value=False,
+    ):
+        await hass.config_entries.async_add(
+            MockConfigEntry(domain=DOMAIN, data=config, unique_id=UNIQUE_ID)
+        )
+        await hass.async_block_till_done()
+        assert len(hass.states.async_entity_ids(MP_DOMAIN)) == 0
+
+
+async def _test_service(
+    hass: HomeAssistantType,
+    vizio_func_name: str,
+    ha_service_name: str,
+    additional_service_data: dict = None,
+    args: Union[str, List[str]] = [],
+    arg_vals: Union[Union[str, int], List[Union[str, int]]] = [],
+) -> None:
+    """Test a generic Vizio media player entity service."""
+    service_data = {ATTR_ENTITY_ID: ENTITY_ID}
+    if additional_service_data:
+        service_data.update(additional_service_data)
+
+    with patch(
+        f"homeassistant.components.vizio.media_player.VizioAsync.{vizio_func_name}"
+    ) as service_call:
+        await hass.services.async_call(
+            MP_DOMAIN, ha_service_name, service_data=service_data, blocking=True,
+        )
+        assert service_call.called
+
+        if not isinstance(args, list):
+            args = [args]
+        if not isinstance(arg_vals, list):
+            arg_vals = [arg_vals]
+
+        if len(args) == len(arg_vals):
+            # For each argument and argument value pair, assert that pair matches what
+            # was passed to function
+            for arg in args:
+                assert (
+                    service_call.call_args[args.index(arg) + 1][arg]
```

Comment:
It's easier to use `unittest.mock.call` to create a call to compare the mock call to.